aten/src/ATen/native/cudnn/MHA.cpp

```diff
@@ -0,0 +1,327 @@
+  memset(&params, 0, sizeof(MHAParams));
+  params.device_id = at::cuda::current_device();
+  params.dataType = fe::DataType_t::HALF;
+  if (q.scalar_type() == kBFloat16) {
+    params.dataType = fe::DataType_t::BFLOAT16;
+  }
+  params.b = b;
+  params.h = h;
+  params.d = d;
+  params.s_q = s_q;
+  params.s_kv = s_kv;
+  params.dropout_probability = dropout_probability;
+  params.is_causal = is_causal;
+  params.return_softmaxstats = return_softmaxstats;
+  TORCH_INTERNAL_ASSERT(q.sizes().size() == MAX_MHA_DIM, "Q tensor has unexpected number of dims, please report a bug to PyTorch.");
+  TORCH_INTERNAL_ASSERT(q.strides().size() == MAX_MHA_DIM, "Q tensor has unexpected number of dims, please report a bug to PyTorch.");
+  TORCH_INTERNAL_ASSERT(k.sizes().size() == MAX_MHA_DIM, "K tensor has unexpected number of dims, please report a bug to PyTorch.");
+  TORCH_INTERNAL_ASSERT(k.strides().size() == MAX_MHA_DIM, "K tensor has unexpected number of dims, please report a bug to PyTorch.");
+  TORCH_INTERNAL_ASSERT(v.sizes().size() == MAX_MHA_DIM, "V tensor has unexpected number of dims, please report a bug to PyTorch.");
+  TORCH_INTERNAL_ASSERT(v.strides().size() == MAX_MHA_DIM, "V tensor has unexpected number of dims, please report a bug to PyTorch.");
+  std::copy(q.sizes().begin(), q.sizes().end(), params.q_dim);
+  std::copy(q.strides().begin(), q.strides().end(), params.q_stride);
+  std::copy(k.sizes().begin(), k.sizes().end(), params.k_dim);
+  std::copy(k.strides().begin(), k.strides().end(), params.k_stride);
+  std::copy(v.sizes().begin(), v.sizes().end(), params.v_dim);
+  std::copy(v.strides().begin(), v.strides().end(), params.v_stride);
+}
+
+struct MHACacheKeyWrapper : ParamsWrapper<MHAParams> {
+  MHACacheKeyWrapper(int64_t b, int64_t h, int64_t s_q, int64_t s_kv, int64_t d, const Tensor& q, const Tensor& k, const Tensor & v, double dropout_probability, bool is_causal, bool return_softmaxstats) {
+    setMHAParams(this->pod, b, h, s_q, s_kv, d, q, k, v, dropout_probability, is_causal, return_softmaxstats);
+  }
+};
+
+template <typename T, typename KeyType>
+struct MHAGraphCache {
+std::unordered_map<KeyType, graph_and_tensors, ParamsWrapperHash<KeyType>> engine_cache;
+
+// no mutexes here as caches are now thread local for v8, can also return a pointer
+// to the Execution Plan if we know it will not be invalidated by another thread
+T* find(const KeyType& key) {
+  auto it = engine_cache.find(key);
+  if (it == engine_cache.end()) {
+    return nullptr;
+  }
+  return &(it->second);
+}
+
+void update(const KeyType& key, T& results) {
+  engine_cache.erase(key);
+  engine_cache.emplace(key, std::move(results));
+}
+
+};
+
+// @eqy: use thread local caches as cuDNN Execution Plans are not guaranteed to be thread safe across all engines
```

Comment:
this comment seems at odds with the one mentioned here: https://github.com/pytorch/pytorch/pull/115663/files#diff-3fdc8929e8ab637660e7f8a73224a364bf57429e920aa4235305a2a0063b82cdR130